language/move-ir-compiler/bytecode-source-map/src/source_map.rs

```diff
@@ -70,17 +74,19 @@ pub struct SourceMap {
     /// The source location for the definition of the module or script that this source map is for.
     pub definition_location: Loc,
 
-    /// The name <address.module_name> for module that this source map is for
-    /// None if it is a script
+    /// The name <address.module_name> of the module that this source map is for.
+    /// `None` if this source map corresponds to a script.
     pub module_name_opt: Option<(AccountAddress, Identifier)>,
 
-    // A mapping of StructDefinitionIndex to source map for each struct/resource
+    // A mapping of `StructDefinitionIndex` to source map for each struct/resource.
     struct_map: BTreeMap<TableIndex, StructSourceMap>,
 
-    // A mapping of FunctionDefinitionIndex to the soure map for that function.
+    // A mapping of `FunctionDefinitionIndex` to the soure map for that function.
+    // For scripts, this map has a single element that points to a source map corresponding to the
+    // script's "main" function.
```

Comment:
What would you think about adding more structure to this field for script v.s. modules (while preserving the APIs)? e.g., something like
```rust
pub enum ScriptOrModuleFnMap {
  Script(FunctionSourceMap),
   Module(BTreeMap<TableIndex, FunctionSourceMap>)
}
```

and then this field would be `function_map: ScriptOrModuleFnMap`

Reply:
I like the idea of adding structure, but JFYI I believe the idea that "scripts have a `main` function at index 0" is baked into several parts of this code, and so a change like this may cascade into a large number of related changes:

1. `SourceMap` methods that build up the mapping between source and bytecode, like `get_code_location` and `add_local_mapping` -- which are used when compiling both scripts and modules -- all take a `FunctionDefinitionIndex` as a parameter. When compiling scripts, a dummy source map that contains a function definition with index 0 is constructed, so these methods "work." But if I change `SourceMap` to either hold onto a single `FunctionSourceMap` for scripts, or a map of indices to `FunctionSourceMap` for modules, then these index parameters will no longer make sense for scripts. I'll need to modify all 9 of these methods to work for the script case.
2. I believe this is how source maps work: empty source maps are first constructed, _then_ the `ir-to-bytecode::compiler` module populates them as it translates Move IR AST into bytecode. This means the top-level `SourceMap` creates an empty map for its `function_map` member. Once the compiler traverses the AST and finds a function definition, a new `FunctionSourceMap` is inserted via a call to `SourceMap::add_top_level_function_mapping`. Constructing even an empty `FunctionSourceMap` requires the function's location, which is known once the function's AST node is reached. However, to construct a `ScriptOrModuleFnMap::Script` with the above definition, I would need to know the function's location up front, before traversing the AST and reaching the script's `main` function node. I don't think that's possible without a lot of changes to how things work right now, so I think this would need to be `ScriptOrModuleFnMap::Script(Option<FunctionSourceMap>)`, or something like that.

It's probably worthwhile to refactor this code now: scripts are now treated as an independent construct during disassembly, instead of being transformed into a "fake" module with one function definition. So, this part of the codebase ought to change as well, to no longer shoehorn function indices into script source mappings. But, just thought I'd mention this, because:

1. I don't want you to be surprised if this PR becomes much larger. (And please do let me know if you think there's a way to use `ScriptOrModuleFnMap` without making this many other changes!)
2. I _think_ I heard someone mention undoing the concept of "scripts" in Move and having them be treated identically to modules. If that's on our roadmap _at all_, then JFYI this change will take us in the exact opposite direction.

Reply:
As we discussed in Slack DM's, I think this might result in a large set of other changes, so I'll leave this be for now and re-(ab)use the function source map index method.